### ISSUE TITLE
derivePersistField phantom type support

### ISSUE THREAD

**Issue by tim2CF** [yesodweb/persistent#1137]
Phantom type are very useful for tagging things on type level, but at the moment this code will not compile

```haskell
-- @Employment.hs
{-# LANGUAGE TemplateHaskell #-}
module Employment where

import Database.Persist.TH
import Prelude

data Employment a = Employed | Unemployed | Retired
    deriving (Show, Read, Eq)
derivePersistField "Employment a"
```

**Comment by parsonsmatt** [comment]
The manual instances shouldn't be too bad to write:

```haskell
instance PersistField (Employment a) where
    toPersistField = toPersistField . show
    fromPersistField pf = do
        str <- fromPersistField
        case readMay str of
            ... -> pure etc
            _ -> fail $ "Couldn't parse Employment, got: " <> str
```

But you're right, this should be supported.

More generally, `derivePersistField` should accept a type name instead of a string, so you'd write `derivePersistField ''Employment`. I wonder why it's a string.

Even better, we ought to have a newtype for `deriving via`...

```haskell
newtype AsShowRead a = AsShowRead a

instance (Show a, Read a) => PersistField (AsShowRead a) where
  toPersistValue (AsShowRead a) = toPersistField (show a)
  fromPersistValue pval = do
    str <- fromPersistValue pval
    case readMay str of
      Just a -> pure (AsShowRead a)
      Nothing -> fail $ "Failed to parse " <> typeName <> " via Read. Got: " <> str
```

Then you can write:

```haskell
data Employment a = ...
  deriving stock (Show, Read, Eq)
  deriving PersistField via AsShowRead (Employment a)
```

This would make a great first issue for anyone to tackle.

**Comment by juanpaucar** [comment]
Hi folks, @parsonsmatt may I get this issue assigned to me, please?

**Comment by parsonsmatt** [comment]
go for it!

**Comment by MaxGabriel** [comment]
If it's helpful, this is a version I wrote at work that uses the exact names of the enum values instead of `Read` instances. This is less general, but significantly faster than `Read`. This is specifically for Postgres enums, but could be adapted for more general usage.

```haskell
module Mercury.Database.Persist.DeriveEnum (derivePostgresEnumForSumType) where

import ClassyPrelude
import Control.Monad.Fail (fail)
import qualified Data.ByteString.Char8 as S8
import qualified Data.Text as T
import Database.Persist (PersistField (..), PersistValue (..))
import Database.Persist.Sql (PersistFieldSql, SqlType (..), sqlType)
import Language.Haskell.TH.Datatype (ConstructorInfo (..), ConstructorVariant (..), DatatypeInfo (..), reifyDatatype)
import Language.Haskell.TH.Lib (varE)
import Language.Haskell.TH.Syntax

-- | Derives PersistField and PersistFieldSql for a sum type, using Postgres enums
--
-- Unlike derivePostgresEnum, this version does not use Show/Read instances, but instead the exact Haskell name
-- By not using Read (or converting from ByteString to String), parsing speed is dramatically increased
-- If necessary, we could probably make this also work for Show instances as well, though.
--
-- Example usage: $(derivePostgresEnumForSumType ''MySumType "my_sum_type")
derivePostgresEnumForSumType :: Name -> String -> Q [Dec]
derivePostgresEnumForSumType typeName postgresType = do
  datatypeInfo <- reifyDatatype typeName
  constructorNames <- mapM getNormalConstructorName (datatypeCons datatypeInfo)

  toPersistValueMatches <- mapM constructorNameToMatch constructorNames
  let toPersistValueExp = LamCaseE toPersistValueMatches

  fromPersistValueGoodMatches <- mapM constructorNameFromMatch constructorNames
  fromPersistValueBadMatches <- fromPersistValueBadMatch typeName
  let fromPersistValueExp = LamCaseE (fromPersistValueGoodMatches <> [fromPersistValueBadMatches])

  sqlTypeImplementation <- [|SqlOther (T.pack postgresType)|]

  return
    [ InstanceD
        Nothing
        []
        (ConT ''PersistField `AppT` ConT typeName)
        [ FunD
            'toPersistValue
            [ Clause [] (NormalB toPersistValueExp) []
            ],
          FunD
            'fromPersistValue
            [ Clause [] (NormalB fromPersistValueExp) []
            ]
        ],
      InstanceD Nothing [] (ConT ''PersistFieldSql `AppT` ConT typeName) [FunD 'sqlType [Clause [WildP] (NormalB sqlTypeImplementation) []]]
    ]

-- | Given a constructor "Foo", generates @Foo -> PersistDbSpecific "Foo"@
constructorNameToMatch :: Name -> Q Match
constructorNameToMatch name = do
  let s = nameToOccNameString name
  anExp <- [|PersistDbSpecific $ S8.pack s|]
  pure $ Match (ConP name []) (NormalB anExp) []

-- | Given a constructor "Foo", generates @"Foo" -> Right Foo@
constructorNameFromMatch :: Name -> Q Match
constructorNameFromMatch name = do
  let s = nameToOccNameString name
  let pat = ConP 'PersistDbSpecific [LitP (StringL s)]

  (rightConstructor :: Exp) <- [|Right|]
  let result = rightConstructor `AppE` ConE name
  pure $ Match pat (NormalB result) []

-- | Generates a fallthrough pattern match for the given type name:
--
-- @unexpectedPersistValue -> Left "fromPersistField: When trying to deserialize a value of type MyEnum, got Baz"@
fromPersistValueBadMatch :: Name -> Q Match
fromPersistValueBadMatch typeName = do
  let s = nameToOccNameString typeName
  unexpectedPersistValueName <- newName "unexpectedPersistValue"
  errorCase <- [|Left $ T.pack $ "fromPersistField: When trying to deserialize a value of type " <> s <> ", got: " <> show $(varE unexpectedPersistValueName)|]
  pure $ Match (VarP unexpectedPersistValueName) (NormalB errorCase) []

getNormalConstructorName :: ConstructorInfo -> Q Name
getNormalConstructorName constructorInfo = case constructorVariant constructorInfo of
  NormalConstructor -> case constructorFields constructorInfo of
    [] -> pure $ constructorName constructorInfo
    bad -> fail $ "derivePostgresEnumForSumType only works for sum types with no arguments (e.g. data X = A | B is OK, data X = A Int | B is not). Got: " <> show bad
  bad -> fail $ "derivePostgresEnumForSumType expects normal (non-record, non-infix) constructors, got: " <> show bad

nameToOccNameString :: Name -> String
nameToOccNameString (Name (OccName s) _) = s

```

**Comment by juanpaucar** [comment]
Thanks @MaxGabriel I'm checking it. I was having a slow start

**Comment by juanpaucar** [comment]
Hi @parsonsmatt. Sorry for the delay I was kind of busy with the job and I was having some hard time getting the template Haskell stuff working. Is there any problem if I take one or two more days?

**Comment by parsonsmatt** [comment]
There is no problem at all :) Feel free to take as much time as you'd like. If you have any questions or are blocked on anything, feel free to ask!

**Comment by tim2CF** [comment]
Hello guys! I found another issue with Persistent TH. I don't know Persistent implementation details, but maybe it's also related https://github.com/yesodweb/persistent/issues/1166
